Mathlib/Dynamics/TopologicalEntropy/NetEntropy.lean

```diff
@@ -0,0 +1,408 @@
+    rw [netEntropyEnt, limsup_congr h]
+    exact limsup_const ⊥
+  simp only [netMaxcard_empty, ENat.toENNReal_zero, log_zero, eventually_atTop]
+  exact ⟨1, fun n n_pos ↦ bot_div_of_pos_ne_top (Nat.cast_pos'.2 n_pos) (natCast_ne_top n)⟩
+
+@[simp]
+lemma netEntropyInfEnt_empty {T : X → X} {U : Set (X × X)} : netEntropyInfEnt T ∅ U = ⊥ :=
+  eq_bot_mono (netEntropyInfEnt_le_netEntropyEnt T ∅ U) netEntropyEnt_empty
+
+lemma netEntropyInfEnt_nonneg (T : X → X) {F : Set X} (h : F.Nonempty) (U : Set (X × X)) :
+    0 ≤ netEntropyInfEnt T F U :=
+  (le_iInf fun n ↦ div_nonneg (log_netMaxcard_nonneg T h U n) (Nat.cast_nonneg' n)).trans
+    iInf_le_liminf
+
+lemma netEntropyEnt_nonneg (T : X → X) {F : Set X} (h : F.Nonempty) (U : Set (X × X)) :
+    0 ≤ netEntropyEnt T F U :=
+  (netEntropyInfEnt_nonneg T h U).trans (netEntropyInfEnt_le_netEntropyEnt T F U)
+
+lemma netEntropyInfEnt_univ (T : X → X) {F : Set X} (h : F.Nonempty) :
+    netEntropyInfEnt T F univ = 0 := by simp [netEntropyInfEnt, netMaxcard_univ T h]
+
+lemma netEntropyEnt_univ (T : X → X) {F : Set X} (h : F.Nonempty) :
+    netEntropyEnt T F univ = 0 := by simp [netEntropyEnt, netMaxcard_univ T h]
+
+lemma netEntropyInfEnt_le_coverEntropyInfEnt (T : X → X) (F : Set X) {U : Set (X × X)}
+    (U_symm : SymmetricRel U) :
+    netEntropyInfEnt T F U ≤ coverEntropyInfEnt T F U := by
+  refine (liminf_le_liminf) (Eventually.of_forall fun n ↦ ?_)
+  apply div_le_div_right_of_nonneg (Nat.cast_nonneg' n) (log_monotone _)
+  exact ENat.toENNReal_le.2 (netMaxcard_le_coverMincard T F U_symm n)
+
+lemma coverEntropyInfEnt_le_netEntropyInfEnt (T : X → X) (F : Set X) {U : Set (X × X)}
+    (U_rfl : idRel ⊆ U) (U_symm : SymmetricRel U) :
+    coverEntropyInfEnt T F (U ○ U) ≤ netEntropyInfEnt T F U := by
+  refine (liminf_le_liminf) (Eventually.of_forall fun n ↦ ?_)
+  apply div_le_div_right_of_nonneg (Nat.cast_nonneg' n) (log_monotone _)
+  exact ENat.toENNReal_le.2 (coverMincard_le_netMaxcard T F U_rfl U_symm n)
+
+lemma netEntropyEnt_le_coverEntropyEnt (T : X → X) (F : Set X) {U : Set (X × X)}
+    (U_symm : SymmetricRel U) :
+    netEntropyEnt T F U ≤ coverEntropyEnt T F U := by
+  refine (limsup_le_limsup) (Eventually.of_forall fun n ↦ ?_)
+  apply div_le_div_right_of_nonneg (Nat.cast_nonneg' n) (log_monotone _)
+  exact ENat.toENNReal_le.2 (netMaxcard_le_coverMincard T F U_symm n)
+
+lemma coverEntropyEnt_le_netEntropyEnt (T : X → X) (F : Set X) {U : Set (X × X)} (U_rfl : idRel ⊆ U)
+    (U_symm : SymmetricRel U) :
+    coverEntropyEnt T F (U ○ U) ≤ netEntropyEnt T F U := by
+  refine (limsup_le_limsup) (Eventually.of_forall fun n ↦ ?_)
+  apply div_le_div_right_of_nonneg (Nat.cast_nonneg' n) (log_monotone _)
+  exact ENat.toENNReal_le.2 (coverMincard_le_netMaxcard T F U_rfl U_symm n)
+
+/-! ### Relationship with entropy via covers -/
+
+variable [UniformSpace X] (T : X → X) (F : Set X)
+
+lemma coverEntropyInf_eq_iSup_netEntropyInfEnt :
```

Comment:
This one and the next one would deserve to be called theorems, and would deserve longer docstrings.

Reply:
Done.